tests/catch/unit/memory/hipFree.cc

```diff
@@ -0,0 +1,580 @@
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+
+#include <hip_test_common.hh>
+#include <hip_test_checkers.hh>
+#include <hip_test_kernels.hh>
+#include "DriverContext.hh"
+
+/*
+ * This testcase verifies [ hipFree || hipFreeArray || hipArrayDestroy || hipHostFree with
+ * hipHostMalloc ]
+ * 1. Check that hipFree implicitly synchronises the device.
+ * 2.Perform multiple allocations and then call hipFree on each pointer concurrently (from unique
+ * threads) for different memory types and different allocation sizes.
+ * 3. Pass nullptr as argument and check that no operation is performed and hipSuccess is returned.
+ * 4. Pass an invalid ptr and check that hipErrorInvalidValue is returned.
+ * 5. Call hipFree twice on the same pointer and check that the implementation handles the second
+ * call correctly.
+ * 6. HipHostFree only:
+ *    Try to free memory that has been registered with hipHostRegister and check that
+ * hipErrorInvalidValue is returned.
+ */
+
+
+enum FreeType { DevFree, ArrayFree, ArrayDestroy, HostFree };
+
+__global__ void waitKernel(clock_t offset) {
+  auto time = clock();
+  while (clock() - time < offset) {
+  }
+}
+
+static __global__ void clock_kernel(clock_t clock_count, size_t* co) {
+  clock_t start_clock = clock();
+  clock_t clock_offset = 0;
+  while (clock_offset < clock_count) {
+    clock_offset = clock() - start_clock;
+  }
+  *co = clock_offset;
+}
+
+// number of clocks the device is running at (device frequency)
+static size_t ticksPerMillisecond = 0;
+
+// helper function used to set the device frequency variable
+
+static size_t findTicks() {
+  hipDeviceProp_t prop;
+  int device;
+  size_t* clockOffset;
+  HIP_CHECK(hipMalloc(&clockOffset, sizeof(size_t)));
+  hipEvent_t start, stop;
+  HIP_CHECK(hipEventCreate(&start));
+  HIP_CHECK(hipEventCreate(&stop));
+  HIP_CHECK(hipGetDevice(&device));
+  HIP_CHECK(hipGetDeviceProperties(&prop, device));
+
+  constexpr float milliseconds = 1000;
+  constexpr float tolerance = 0.00002 * milliseconds;
+
+  clock_t devFreq = static_cast<clock_t>(prop.clockRate);  // in kHz
+  clock_t time = devFreq * milliseconds;
+  // Warmup
+  hipLaunchKernelGGL(clock_kernel, dim3(1), dim3(1), 0, 0, time, clockOffset);
+  HIP_CHECK(hipGetLastError());
+  HIP_CHECK(hipDeviceSynchronize());
+
+  // try 10 times to find device frequency
+  int attempts = 6;
+  size_t co = 0;
+
+  while (attempts > 0) {
```

Comment:
use a for-loop here to keep the decrementing simpler

Reply:
While replaced with for loop